roxie/ccd/ccdqueue.cpp

```diff
@@ -294,27 +317,76 @@ void closeMulticastSockets()
     multicastSocket.clear();
 }
 
-size32_t channelWrite(unsigned channel, void const* buf, size32_t size)
+static bool channelWrite(RoxiePacketHeader &buf, bool includeSelf)
 {
     size32_t minwrote = 0;
     if (roxieMulticastEnabled)
     {
-        return multicastSocket->udp_write_to(multicastEndpoints.item(channel), buf, size);
+        return multicastSocket->udp_write_to(multicastEndpoints.item(buf.channel), &buf, buf.packetlength) == buf.packetlength;
     }
     else
     {
+#ifdef _CONTAINERIZED
+        // In the containerized system, the list of subchannel IPs is captured in the packet header to ensure everyone is using the
+        // same snapshot of the topology state.
+        // If the subchannel IPs are not set, fill them in now. If they are set, use them.
+        if (buf.subChannels[0].isNull())
+        {
+            Owned<const ITopologyServer> topo = getTopology();
+            const SocketEndpointArray &eps = topo->queryAgents(buf.channel);
+            if (!eps.ordinality())
+                throw makeStringExceptionV(0, "No agents available for channel %d", buf.channel);
+
+            unsigned hdrHashVal = buf.priorityHash();
+            unsigned numAgents = eps.ordinality();
+            unsigned subChannel = (hdrHashVal % numAgents);
+
+            for (unsigned idx = 0; idx < MAX_SUBCHANNEL; idx++)
+            {
+                if (idx == numAgents)
+                    break;
+                buf.subChannels[idx].setIp(eps.item(subChannel));
+                subChannel++;
+                if (subChannel == numAgents)
+                    subChannel = 0;
+            }
+        }
+        for (unsigned subChannel = 0; subChannel < MAX_SUBCHANNEL; subChannel++)
+        {
+            if (buf.subChannels[subChannel].isNull())
+                break;
+            if (includeSelf || !buf.subChannels[subChannel].isMe())
+            {
+                if (traceRoxiePackets)
+                {
+                    StringBuffer s, header;
+                    DBGLOG("Writing %d bytes to subchannel %d (%s) %s", buf.packetlength, subChannel, buf.subChannels[subChannel].getTraceText(s).str(), buf.toString(header).str());
+                }
+                SocketEndpoint ep(ccdMulticastPort, buf.subChannels[subChannel].getIpAddress());
+                size32_t wrote = multicastSocket->udp_write_to(ep, &buf, buf.packetlength);
+                if (!subChannel || wrote < minwrote)
+                    minwrote = wrote;
+            }
+            else if (traceRoxiePackets)
+            {
+                StringBuffer s, header;
+                DBGLOG("NOT writing %d bytes to subchannel %d (%s) %s", buf.packetlength, subChannel, buf.subChannels[subChannel].getTraceText(s).str(), buf.toString(header).str());
```

Comment:
if there is a single channel, minwrote will still be 0.  will that cause problems since the function will now return false.

Reply:
It would cause an additional line of tracing (IBYTI wrote too little) - that's the only place where includeSelf is passed in false. I can avoid calling altogether in such cases.